community/kernel/src/main/java/org/neo4j/kernel/impl/newapi/Read.java

```diff
@@ -49,36 +57,73 @@
 import org.neo4j.values.storable.Value;
 import org.neo4j.values.storable.ValueGroup;
 
+import static java.lang.String.format;
 import static org.neo4j.kernel.impl.newapi.References.clearFlags;
 import static org.neo4j.kernel.impl.newapi.References.hasDirectFlag;
 import static org.neo4j.kernel.impl.newapi.References.hasFilterFlag;
 import static org.neo4j.kernel.impl.newapi.References.hasGroupFlag;
 import static org.neo4j.kernel.impl.store.record.AbstractBaseRecord.NO_ID;
+import static org.neo4j.unsafe.impl.internal.dragons.FeatureToggles.flag;
+import static org.neo4j.unsafe.impl.internal.dragons.FeatureToggles.toggle;
 
 abstract class Read implements TxStateHolder,
         org.neo4j.internal.kernel.api.Read,
         org.neo4j.internal.kernel.api.ExplicitIndexRead,
-        org.neo4j.internal.kernel.api.SchemaRead
+        org.neo4j.internal.kernel.api.SchemaRead,
+        CursorTracker
 {
     private final Cursors cursors;
     final KernelTransactionImplementation ktx;
 
+    private long cursorsInUse;
+
+    private static final boolean TRACK_CURSORS = flag( Read.class, "trackCursors", false );
+    private static final boolean RECORD_CURSORS_TRACES = flag( KernelStatement.class, "recordCursorTraces", false );
+    private static final int CURSORS_TRACK_HISTORY_MAX_SIZE = 100;
+    private static final LruCache<Cursor, StackTraceElement[]>  EMPTY_CURSOR_HISTORY = new LruCache<>( "EmptyLru", 1 );
+    private final LruCache<Cursor, StackTraceElement[]> cursorsOpenCloseCalls;
+
     Read( Cursors cursors, KernelTransactionImplementation ktx )
     {
         this.cursors = cursors;
         this.ktx = ktx;
+        this.cursorsOpenCloseCalls = RECORD_CURSORS_TRACES ?
+                                     new LruCache<>( "LruCursors", CURSORS_TRACK_HISTORY_MAX_SIZE ) : EMPTY_CURSOR_HISTORY;
     }
 
     @Override
-    public final void nodeIndexSeek(
-            org.neo4j.internal.kernel.api.IndexReference index,
+    public void acquireCursor( Cursor cursor )
+    {
+        if ( TRACK_CURSORS )
+        {
+            StackTraceElement[] stackTrace = Thread.currentThread().getStackTrace();
+            cursorsOpenCloseCalls.put( cursor, Arrays.copyOfRange(stackTrace, 2, stackTrace.length) );
```

Comment:
Looks like setting `RECORD_CURSOR_TRACES` to `false` will increase the overhead of recording cursor traces, since the traces are captured regardless, and then stored in a cache shared among all threads, and the cache has to evict on all but the first put.

Reply:
Also, setting `TRACK_CURSORS` to `true` but leaving `RECORD_CURSOR_TRACES` at `false` will only keep a single element of history, which can help with finding bugs but not so much with finding fixes.

Reply:
So you propose we set the lruCache size in either way to `CURSORS_TRACK_HISTORY_MAX_SIZE`?

Reply:
Yeah, and then you only need a single feature toggle for the lot.